src/core/or/circuitpadding.c

```diff
@@ -851,12 +896,59 @@ circpad_machine_remove_token(circpad_machine_runtime_t *mi)
     mi->nonpadding_sent /= 2;
   }
 
+  /* Update any state packet length limits that apply */
+  circpad_machine_update_state_length(mi);
+
+  /* Remove a token from the histrogram, if applicable */
+  circpad_machine_remove_token(mi);
+}
+
+static inline void
+circpad_machine_update_state_length(circpad_machine_runtime_t *mi)
+{
+  const circpad_state_t *state = NULL;
+
+  if (mi->state_length != CIRCPAD_STATE_LENGTH_INFINITE)
```

Comment:
Is this test backwards?  In other places, INFINITE means "don't decrement it."  But here we decrement it only if it is infinite.

Should we be worried that the tests didn't catch this?

Reply:
Holy moly. Yes and a little. Our tests never tested state lengths that also wanted to count non-padding cells as part of the length of the state. Dag.

I will write a test that fails for this case and then fix this.

Reply:
Nice find! Here is the new test: de62ccd567fb0536bef5c66f8704f4c44086fbde

Here are two typo bugs it found (yours and one other): 0a10fc6208d7e93f1cc8cac0299ee7e78dd37b3c

And here is a token empty bug it found if a padding packet causes us to run out of tokens: ae719d348e1253ad126a0e57a85bdb7c85a11c70